pkg/api/pin.go

```diff
@@ -0,0 +1,167 @@
+// Copyright 2021 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package api
+
+import (
+	"errors"
+	"net/http"
+
+	"github.com/ethersphere/bee/pkg/jsonhttp"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/gorilla/mux"
+)
+
+// pinChunk pin's the already created chunk given its address.
+// It fails if the chunk is not present in the local store.
+// It also increments a pin counter to keep track of how many
+// pin requests are originating for this chunk.
+func (s *server) pinChunk(w http.ResponseWriter, r *http.Request) {
+	addr, err := swarm.ParseHexAddress(mux.Vars(r)["address"])
+	if err != nil {
+		s.logger.Debugf("pin chunk: unable to parse chunk address: %v", err)
+		s.logger.Error("pin chunk: unable to parse address")
+		jsonhttp.BadRequest(w, "bad address")
+		return
+	}
+
+	switch err := s.storer.Set(r.Context(), storage.ModeSetPin, addr); {
+	case errors.Is(err, storage.ErrNotFound):
+		ch, err := s.storer.Get(r.Context(), storage.ModeGetRequest, addr)
+		if err != nil {
+			s.logger.Debugf("pin chunk: unable to get chunk for address %q from netstore: %v", addr, err)
+			s.logger.Error("pin chunk: unable to get chunk for address %q from netstore")
+			jsonhttp.NotFound(w, nil)
+			return
+		}
+
+		_, err = s.storer.Put(r.Context(), storage.ModePutRequestPin, ch)
```

Comment:
the getrequest already saves this to localstore, in fact the put will come back after a has call returns true